crates/typst-pdf/src/page.rs

```diff
@@ -718,21 +754,120 @@ fn write_group(ctx: &mut PageContext, pos: Point, group: &GroupItem) {
+    // example) or normal text we can render it directly
+    let has_color_glyphs = tables.sbix.is_some()
+        || tables.cbdt.is_some()
+        || tables.svg.is_some()
+        || tables.colr.is_some();
+    if !has_color_glyphs {
+        write_normal_text(ctx, pos, TextItemView::all_of(text));
+        return;
+    };
+
+    let color_glyph_count =
+        text.glyphs.iter().filter(|g| is_color_glyph(&text.font, g)).count();
+
+    if color_glyph_count == text.glyphs.len() {
+        write_color_glyphs(ctx, pos, TextItemView::all_of(text));
+    } else if color_glyph_count == 0 {
+        write_normal_text(ctx, pos, TextItemView::all_of(text));
+    } else {
+        // Otherwise we need to split it in smaller text runs
+        let mut offset = 0;
+        let mut position_in_run = Abs::zero();
+        while offset < text.glyphs.len() {
+            // Start a new text run where the last one ended
+            let start = offset;
+            // Determine if this is a color-only or a text-only run
+            let in_colored_group = is_color_glyph(&text.font, &text.glyphs[start]);
+            // Determine the index of the last glyph of the run
+            let end = start
+                + text.glyphs[start..]
+                    .iter()
+                    .position(|g| is_color_glyph(&text.font, g) != in_colored_group)
+                    .unwrap_or(text.glyphs.len() - start);
+
+            // Build a sub text-run
+            let text_item_view = TextItemView::from_glyph_range(text, start..end);
+
+            // Adjust the position of the run on the line
+            let pos = pos + Point::new(position_in_run, Abs::zero());
+            position_in_run += text_item_view.width();
+            offset = end;
+            // Actually write the sub text-run
+            if in_colored_group {
+                write_color_glyphs(ctx, pos, text_item_view);
+            } else {
+                write_normal_text(ctx, pos, text_item_view);
+            }
+        }
+    }
+}
+
+// Encodes a text run made only of color glyphs into the content stream
+fn write_color_glyphs(ctx: &mut PageContext, pos: Point, text: TextItemView) {
+    let x = pos.x.to_f32();
+    let y = pos.y.to_f32();
+
+    let mut last_font = None;
+
+    ctx.content.begin_text();
+    ctx.content.set_text_matrix([1.0, 0.0, 0.0, -1.0, x, y]);
+    // so that the next call to ctx.set_font() will change the font
+    // one that displays regular glyphs and not color glyphs
+    ctx.state.font = None;
+
+    for glyph in text.glyphs() {
+        // artificially choose better resolutions of color glyphs, as they tend
+        // to appear pixelated even at low zoom levels otherwise
+        let ppem = 2.0 * text.item.size.to_f32() as f64;
+        let (font, index) = ctx.parent.color_font_map.get(
+            &mut ctx.parent.alloc,
+            &text.item.font,
+            glyph.id,
+            ppem,
+            || color::frame_for_glyph(&text, &glyph),
```

Comment:
Since this will be reused on a font + glyph ID basis, all other data from `&text` used by `frame_for_glyph`, will be sort of arbitrary based on whichever glyph came first, won't it?

Reply:
The only other data that `frame_for_glyph` uses is the fill color in case of COLRv0 glyphs without a specified fill, which should not occur very often (I'm not even sure Segoe UI Emoji contains any such glyph).
However, if you feel like it should be taken into account too, I can add it to the "key" used for de-duplication.